doc: vpc name must be known before applying the changes.

imprateeksh opened the pull request:
### Description

Updated the variable name which is used only to name the VPE services/gateways.

Reference Issue: #559

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [x] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content
* updated variable descriptions for better code readability.
 
<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [x] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

imprateeksh commented:
/run pipeline

imprateeksh commented:
/run pipeline

imprateeksh commented:
In the second run, `TestRunEveryMultiTenantExample` passed but other two tests failed.
Tried the failed tests locally - advanced test is passing - 
<img width="1379" alt="image" src="https://github.com/user-attachments/assets/bfebdde0-2b7f-4cbc-8f3a-bd93843b663c">
[TestRunAdvancedExample](https://github.com/user-attachments/files/16560342/output_test_3.txt)

But upgrade test is failing in local run as well.
[TestRunUpgradeExample](https://github.com/user-attachments/files/16560352/output_test_4.txt)

imprateeksh commented:
/run pipeline

imprateeksh commented:
/run pipeline

imprateeksh commented:
Advanced Example and Upgrade tests failure are seen because of Postgresql db changes.

imprateeksh commented:
/run pipeline

imprateeksh commented:
After updating the module versions, the tests have [passed](https://github.com/terraform-ibm-modules/terraform-ibm-vpe-gateway/actions/runs/10349848876/job/28645301123) except upgrade tests.

@shemau @toddgiguere - Will it be okay if I skip the upgrade test ?

imprateeksh commented:
Will be skipping the upgrade tests for now because the Upgrade Test is failing. The failure is not due to the base module but the dependent postgresql module has received multiple changes. 
In the base module, postgres module consumed version is [v3.14.3](https://github.com/terraform-ibm-modules/terraform-ibm-vpe-gateway/blob/7542833fccab38151574751940a7b55311d0402b/examples/advanced/main.tf#L68) and the latest postgres module version (as of now) is [v3.16.3](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/releases/tag/v3.16.3)
`updated default value of `member_memory_mb` and inclusion of member_host_flavor` - This is likely the reason for the upgrade failure.

Hence it is safe to skip the upgrade tests.

It appea

imprateeksh commented:
/run pipeline

imprateeksh commented:
`terraform validate` is not failing in local but in pipeline
![image](https://github.com/user-attachments/assets/56aaf23f-7cca-4261-b654-7c1035de8817)

imprateeksh commented:
/run pipeline

toddgiguere commented on variables.tf:
```diff
@@ -15,7 +15,7 @@ variable "prefix" {
 
 variable "vpc_name" {
-  description = "Name of the VPC where the Endpoint Gateways will be created. This value is used to dynamically generate VPE names."
+  description = "Name of the VPC where the Endpoint Gateways will be created. This value is used to generate VPE names dynamically and must be specified before applying the changes."
```
I don't think this description is going to make it anymore clear what the issue really is. The primary issue is that this variable:
1. Is used ONLY to generate VPE names (so that first part that says "Name of the VPC where" is misleading)
2. Whatever string is provided here MUST be a value that is known to terraform before ANY deployment has occurred, such as a static string or an input variable from the user

imprateeksh replied:
Hi @toddgiguere,
Thank you for you comments. Just need to make sure of one thing that apart from the change in variable description, do you suggest to make some change in the logic to first verify whether the VPC provided is correct/already exists otherwise error out?
```hcl
data ibm_is_vpc get_vpc {
    name = var.vpc_name
}
```
Getting below output for incorrect vpc/ vpc that does not exists :

![image](https://github.com/user-attachments/assets/660c0183-f577-4367-97e1-e9c19d34c1f7)

toddgiguere replied:
@imprateeksh No we do not need to validate the value of that name, it is only used as a prefix to create the VPE name and can actually be any value.

I don't see that code block above, `data.ibm_is_vpc.get_vpc`, anywhere in this module or examples?

toddgiguere replied:
Added @SirSpidey as reviewer, he may be able to help with the wording.

Here is initial issue: https://github.com/terraform-ibm-modules/terraform-ibm-vpe-gateway/issues/559

imprateeksh replied:
Sure thank you @toddgiguere , this clarifies. I used `data.ibm_is_vpc.get_vpc` in a separate example to ask you whether this is what we need to include. Now this is clear.

SirSpidey commented:
@toddgiguere @imprateeksh Some questions to help me understand this variable:

- The original description says that this value identifies the VPC where the VPEs are created. Is that correct? If so, let's use that. The new description, "used to created the names of the VPEs..." is vague to me. _How_ is it used? 
- Does the `prefix` variable come into play with `VPC_name`

### Possible description 
The name of the VPC where the virtual private endpoint gateways are created. The VPC must exist before you can deploy this module.

---

> The primary issue is that this variable:
> - Is used ONLY to generate VPE names (so that first part that says "Name of the VPC where" is misleading)
> - Whatever string is provided here MUST be a value that is known to terraform before ANY deployment has occurred, such as a static string or an input variable from the user

imprateeksh commented:
> @toddgiguere @imprateeksh Some questions to help me understand this variable:
> 
>     * The original description says that this value identifies the VPC where the VPEs are created. Is that correct? If so, let's use that. The new description, "used to created the names of the VPEs..." is vague to me. _How_ is it used?
> 
>     * Does the `prefix` variable come into play with `VPC_name`
> 
> 
> ### Possible description
> 
> The name of the VPC where the virtual private endpoint gateways are created. The VPC must exist before you can deploy this module.
> 
> > The primary issue is that this variable:
> > 
> > * Is used ONLY to generate VPE names (so that first part that says "Name of the VPC where" is misleading)
> > * Whatever string is provided here MUST be a value that is known to terraform before ANY deployment has occurred, such as a static string or an input variable from the user

Hi @SirSpidey Thank you for your feedback. I mentioned _"used to created the name.._" as the `vpc_name` is used to construct name of the VPE as it is getting appended with the VPE name.
Also the variable `prefix` is not used with `vpc_name` but is used with the `vpe` name. 
For example: 
```hcl
name = service.vpe_name != null ? service.vpe_name : "${var.prefix}-${var.vpc_name}-${service.service_name}"
# Constructing name using prefix and vpc which must already exist i.e. "${var.prefix}-${var.vpc_name}-${service.service_name}"
```
I have modified the variable name as per your suggestion.

SirSpidey commented:
@imprateeksh Thanks for the details on the vpe name. It looks like my possible description isn't quite right. 

Is this then accurate?

```markdown
The name of the VPC that's added to the service name of the virtual private endpoints. This VPC must exist before you can deploy this module. If a `prefix` input variable is specified, it's also added to the name of the VPE in the `<prefix>-<vpc_name>-` format
```

imprateeksh commented:
> ```
> The name of the VPC that's added to the service name of the virtual private endpoints. This VPC must exist before you can deploy this module. If a `prefix` input variable is specified, it's also added to the name of the VPE in the `<prefix>-<vpc_name>-` format
> ```
Sure, I think this is fine but can we avoid mentioning `prefix` here since this is about variable `vpc_name` description, and including the prefix might not be necessary. What is your suggestion?

imprateeksh commented:
/run pipeline

SirSpidey commented:
> Sure, I think this is fine but can we avoid mentioning `prefix` here since this is about variable `vpc_name` description, and including the prefix might not be necessary. What is your suggestion?

Sorry, I misread your earlier note. I guess this does it:

```md
The name of the VPC that's used as part of virtual private endpoint names. This VPC must exist before you can deploy this module.
```

---

I might be obsessing, but it looks like users in the console UI wouldn't see the info about the vpe_name because it's not in a description. Maybe we need to add the information in the `cloud_services` and `cloud_service_by_crn` variables so that it's visible there. 

`cloud_services` description:
```md
"The list of cloud services used to create endpoint gateways. If `vpe_name` is not specified in the list, VPE names are created as `<prefix>-<vpc_name>-<service_name>`.
```

error_message for `cloud_services`:
```md
"The service you're trying to add is not supported. For a list of supported services, see [VPE-enabled services](https://cloud.ibm.com/docs/vpc?topic=vpc-vpe-supported-services). You can add unsupported services in the `cloud_service_by_crn` variable."
```

`cloud_service_by_crn` description:
```md
"The list of cloud service CRNs used to create endpoint gateways. Use this list to identify services that are not supported by service name in the `cloud_services` variable. For a list of supported services, see [VPE-enabled services](https://cloud.ibm.com/docs/vpc?topic=vpc-vpe-supported-services). If `service_name` is not specified, the CRN is used to find the name. If `vpe_name` is not specified in the list, VPE names are created as `<prefix>-<vpc_name>-<service_name>`
```

imprateeksh commented:
/run pipeline

imprateeksh commented:
/run pipeline

imprateeksh commented:
Advanced Example had failed in pipeline, updated the postgres version and it passed locally. Here are the logs
[advanced_eg.txt](https://github.com/user-attachments/files/16705637/advanced_eg.txt)
Updated the version and re-running the pipeline again.

toddgiguere commented:
@SirSpidey @imprateeksh this is still probably not correct:
> The name of the VPC that's used as part of virtual private endpoint names. This VPC must exist before you can deploy this module.

This is where the variable name is not really great but I don't think we want to change it.

The `vpc_name` value does NOT need to be a real vpc name, belonging to a real VPC. It is just used to name a new VPE if a name is not supplied, so it can really be any value that the user wants to use to differentiate the VPE name.
For instance:
I have a VPC called "my-app-vpc-for-management-00834". I could pass the string "managment" for the `vpc_name` into this module to get a VPE name of "the-prefix-management-theservice".

It is not "This VPC must exist", it is more of an issue about the `vpc_name` itself must be a string that is known before anything is applied. 

Now that I see the description of cloud_services, I wonder if we should do something similar for `vpc_name`, saying that it is a value used in creating VPE names if names are not supplied, maybe even give the format, and then add that it must be a string that is known before apply?

SirSpidey commented:
@toddgiguere @imprateeksh 
> The vpc_name value does NOT need to be a real vpc name, belonging to a real VPC.

> Now that I see the description of cloud_services, I wonder if we should do something similar for `vpc_name`, saying that it is a value used in creating VPE names if names are not supplied, maybe even give the format, and then add that it must be a string that is known before apply?

Ok. Let's remove the _must exist_ statement. And I did have similar info in an earlier https://github.com/terraform-ibm-modules/terraform-ibm-vpe-gateway/pull/561#issuecomment-2298892440, but perhaps this works better. The format isn't here because it's there where it really applies -- in the list description: 

**Description of vpc_name**
_A label that's used to name virtual private endpoints if `vpe_name` is not specified in the `cloud_services` or `cloud_service_by_crn` input variable._ 


> ... and add that it must be a string that is known before apply?

I'm not sure what you mean by that. How does this figure in the cloud console UI?

toddgiguere commented:
@SirSpidey 
> ... and add that it must be a string that is known before apply?

This is the main reason we wanted to change the description, because the value of `vpc_name` must be a value that terraform has BEFORE the first `apply` (in other words, during `plan` time).

For instance, this will work:
`vpc_name = "${var.prefix}-myname"` because terraform knows what this value is before anything is deployed.

This would NOT work:
`vpc_name = "${var.prefix}-${ibm_is_vpc.guid}"`
because the value of `ibm_is_vpc.guid` is not known until AFTER apply, and in this module doing something like that will cause an error at plan time.

toddgiguere commented:
@SirSpidey what I mean by referencing cloud_services was the structure of the description, not that they are related.

So maybe something like:
```
A label that's used to name virtual private endpoints. If `vpe_name` is not specified in the list, VPE names are created as `<prefix>-<vpc_name>-<service_name>`.
```
and then add to the end of that something about "The value supplied must be something known at `plan` time."  ??? something like that.

toddgiguere commented:
@SirSpidey my suggestion above was ONLY for the `vpc_name` description, trying to not imply it had to be a real vpc name.

The reason we need to add something about it "being known at plan time" was actually the cause for the creation of this issue. We had a consumer of this module who had a very complex setup for VPCs (extremely dynamic) and they had a `vpc_name` input that would only resolve after another resource was deployed, and when they got this error they were confused why.

SirSpidey commented:
I'm stumped, @toddgiguere. If we refer to `plan time`, we'll likely confuse the cloud projects users. And I'm not clear how those projects users could have a dynamic `vpc_name` value unless it's through a reference to another stack member. 

That said, what about adding, "If you specify a value by reference to another configuration, make sure that the value is available when this module is deployed." ???

SirSpidey commented:
@toddgiguere @ocofaigh @imprateeksh How about this? 
> A label that can be used as a short name for virtual private endpoints. If `vpe_name` is not specified in the `cloud_services` or `cloud_service_by_crn` input variable lists, VPE names are created in the format `<prefix>-<vpc_name>-<service_name>`. The value that you specify for `vpc_name` must be known at Terraform `plan` time.

imprateeksh commented:
/run pipeline

imprateeksh commented:
/run pipeline

SirSpidey commented on variables.tf:
```diff
@@ -15,7 +15,7 @@ variable "prefix" {
 
 variable "vpc_name" {
-  description = "Name of the VPC where the Endpoint Gateways will be created. This value is used to dynamically generate VPE names."
+  description = "A label that can be used as a short name for virtual private endpoints. If `vpe_name` is not specified in the `cloud_services` or `cloud_service_by_crn` input variable lists, VPE names are created in the format `<prefix>-<vpc_name>-<service_name>`. The value that you specify for `vpc_name` must be known at Terraform plan time"
```
```suggestion
  description = "A label that can be used as a short name for virtual private endpoints. If `vpe_name` is not specified in the `cloud_services` or `cloud_service_by_crn` input variable lists, VPE names are created in the format `<prefix>-<vpc_name>-<service_name>`. The value that you specify for `vpc_name` must be known at Terraform plan time."
```

imprateeksh replied:
@toddgiguere @ocofaigh - Can you also please suggest if this looks ok and the changes can be merged?

ocofaigh replied:
@imprateeksh suggestion looks good here

imprateeksh replied:
sure, I have made this change. Thank you Conall and Allen.

SirSpidey reviewed (dismissed):
These look good from my perspective. Just need other reviewers to agree that these descriptions address the original issues.

imprateeksh commented:
/run pipeline

ocofaigh commented:
@imprateeksh Lets commit the suggestion in https://github.com/terraform-ibm-modules/terraform-ibm-vpe-gateway/pull/561#discussion_r1738532319 and we can merge. Ill generate a patch release since this is updating the module inputs

imprateeksh commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 4.3.2 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-vpe-gateway/releases/tag/v4.3.2)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
